Allowing n_params > n_dims in function solve()

piwanek opened the pull request:
For n_params > n_dims we (me and @rpoleski) have tested that this solution works.

JohannesBuchner commented:
Thanks. Could you please add a unit test? see https://github.com/JohannesBuchner/PyMultiNest/blob/master/tests/pymultinest_test.py 
The existing tests could be more sophisticated, at the moment they just test that nothing is crashing, although pymultinest_minimal_test.py verifies that the functions are called.

rpoleski commented:
I think this change also addresses the issue #108 raised by @jluastro.

piwanek commented:
@JohannesBuchner, requested unit test was added.

JohannesBuchner commented on tests/pymultinest_test.py:
```diff
@@ -38,6 +38,42 @@ def myloglike(cube):
 	assert numpy.isclose(result['logZ'], 236.0, atol=1), result['logZ']
 	assert numpy.isclose(result['logZerr'], 0.1, atol=1), result['logZerr']
 
+def test_2():
```
add here 
	test_2.cube_dim = None
	test_2.params_dim = None

JohannesBuchner commented on tests/pymultinest_test.py:
```diff
@@ -38,6 +38,42 @@ def myloglike(cube):
+		cube[-1] = numpy.sum(cube[:-1])
+		return cube * 10 * pi
+
+	def myloglike(cube):
```
add here:
	test_2.params_dim = cube.shape

JohannesBuchner commented on tests/pymultinest_test.py:
```diff
@@ -38,6 +38,42 @@ def myloglike(cube):
 
+def test_2():
+	"""testing n_params != n_dims case"""
+	def myprior(cube):
```
add here:
	test_2.prior_dim = cube.shape

JohannesBuchner commented on tests/pymultinest_test.py:
```diff
@@ -38,6 +38,42 @@ def myloglike(cube):
+	print('parameter values:')
+	for name, col in zip(parameters, result['samples'].transpose()):
+		print('%15s : %.3f +- %.3f' % (name, col.mean(), col.std()))
+
```
verify here with assert that
test_2.prior_dim == (n_dims,)
test_2.params_dim == (n_params,)

JohannesBuchner commented on tests/pymultinest_test.py:
```diff
@@ -38,6 +38,42 @@ def myloglike(cube):
+	assert numpy.isclose(result['logZ'], 239.1, atol=1), result['logZ']
+	assert numpy.isclose(result['logZerr'], 0.1, atol=1), result['logZerr']
+
+
```
and please do the same in test_1

piwanek commented:
It turns out that prior function gets array of size (n_params,) and likelihood function gets array of size (n_dims,). We corrected chi2 calculation: `cos(cube[:-1] / 2.)` => `cos(cube / 2.)`. Now the evidence test results in the same value. Thanks for these suggestions. First unit test was also modified.

rpoleski commented:
It seems that all the suggestions by @JohannesBuchner were incorporated. Now, this pull request also improves previously existing unit test.
